apps/bot/src/services/twitch-irc/index.ts

```diff
@@ -1,45 +1,83 @@
-import { TTwitchIrcContext } from "services/types";
-import { TOption } from "types";
+import type { TTwitchIrcContext } from "services/types";
+import type { TOption } from "types";
 import { logger } from "utils/logger";
-import { MessageEvent, WebSocket } from "ws";
+import { type MessageEvent, WebSocket } from "ws";
 import { parseMessage } from "./parsers";
+import { storage } from "services/db";
+
+// NOTE: I think 10 minutes should be safe time
+const PING_TRESHHOLD = 600_000;
+
+type TResType = (value: TOption<boolean> | PromiseLike<TOption<boolean>>) => void;
 
 export class TwitchIrc {
   private ws: WebSocket;
-  private nick: string;
-  private password: string;
   private handlers: Map<string, (ctx: TTwitchIrcContext) => void> = new Map();
 
-  constructor(url: string, nick: string, password: string) {
-    this.nick = nick;
-    this.password = password;
+  private static _instance: TwitchIrc | undefined;
+
+  private constructor(
+    private url: string,
+    private nick: string,
+    private password: string,
+  ) {
     this.ws = new WebSocket(url);
   }
 
-  public async connect(): Promise<TOption<TwitchIrc>> {
-    return new Promise((res) => {
+  public static instance(url: string, nick: string, password: string) {
+    if (!TwitchIrc._instance) {
+      TwitchIrc._instance = new TwitchIrc(url, nick, password);
+      return TwitchIrc._instance;
+    }
+
+    return TwitchIrc._instance;
+  }
+
+  public startPingCheck() {
+    setInterval(() => {
+      const lastping = Number(storage.get("lastping")?.value);
+      const now = Date.now();
+      if (now - lastping > PING_TRESHHOLD) {
+        this.reconnect();
+      }
+    }, 60_000);
+  }
+
+  public async connect(): Promise<void> {
+    await new Promise((resolve: TResType) => {
```

Comment:
~~It should return this `Promise` right?~~. 
~~Otherwise you are evaluating the a `async connect` or did I got something wrong?~~   
(feels like I'm too much on Go now)

I'll leave this comment to see my line of thought, I didn't notice the `await` before the `new Promise` xD